invenio_records_resources/services/custom_fields/date.py

```diff
@@ -0,0 +1,101 @@
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Custom Fields for InvenioRDM."""
+
+import calendar
+
+from arrow import Arrow
+from edtf import parse_edtf
+from edtf.parser.edtf_exceptions import EDTFParseException
+from marshmallow import fields
+from marshmallow_utils.fields import EDTFDateString, ISODateString
+
+from .base import BaseListCF
+
+
+class ISODateStringCF(BaseListCF):
+    """ISO date custom field."""
+
+    @property
+    def mapping(self):
+        """Return the mapping."""
+        return {"type": "date"}
+
+    @property
+    def field(self):
+        """Marshmallow field custom fields."""
+        _schema = ISODateString(**self._field_args)
+        if self._multiple:
+            return fields.List(_schema)
+        return _schema
+
+
+class EDTFDateStringCF(BaseListCF):
+    """EDTF date custom field."""
+
+    @property
+    def mapping(self):
+        """Return the mapping."""
+        return {
+            "type": "object",
+            "properties": {
+                "date": {"type": "date"},
+                "date_range": {"type": "date_range"},
+            },
+        }
+
+    @property
+    def field(self):
+        """Marshmallow field custom fields."""
+        _schema = EDTFDateString(**self._field_args)
+        if self._multiple:
+            return fields.List(_schema)
+        return _schema
+
+    @classmethod
+    def _format_date(cls, date):
+        """Format the given date into ISO format."""
+        arrow = Arrow.fromtimestamp(calendar.timegm(date), tzinfo=utc)
+        return arrow.date().isoformat()
+
+    @classmethod
+    def _calculate_date_range(cls, date):
+        pd = parse_edtf(date)
+        return {
+            "date": date,
+            "date_range": {
+                "gte": cls._format_date(pd.lower_strict()),
+                "lte": cls._format_date(pd.upper_strict()),
+            },
+        }
+
+    def dump(self, record, cf_key="custom_fields"):
+        """Dump the custom field.
+
+        Gets both the record and the custom fields key as parameters.
+        This supports the case where a field is based on others, both
+        custom and non-custom fields.
+        """
+        dates = record[cf_key].get(self.name)
+        if dates:
+            if self._multiple:
+                for date in dates:
+                    record[cf_key][self.name] = self._calculate_date_range(date)
+            else:
+                record[cf_key][self.name] = self._calculate_date_range(date)
+
+    def load(self, record, cf_key="custom_fields"):
+        """Load the custom field.
+
+        Gets both the record and the custom fields key as parameters.
+        This supports the case where a field is based on others, both
+        custom and non-custom fields.
+        """
+        date = record.get(cf_key, {}).pop(self.name, None)
```

Comment:
Not sure why, but on search a record passes twice by the dumper, the second time without `metadata` and `custom_fields`.  That's why the first is a `.get()` instead of `[]`
```shell
(Pdb) record.keys()
dict_keys(['parent', 'versions', 'has_draft', '$schema', 'pid', 'id', 'files', 'metadata', 'custom_fields', 'pids', 'access', 'uuid', 'version_id', 'created', 'updated'])
(Pdb) record.keys()
dict_keys(['id', 'pid', 'access', '$schema', 'communities', 'uuid', 'version_id', 'created', 'updated'])
```